pkg/domains/trace/trace.go

```diff
@@ -0,0 +1,174 @@
+//
+// # Class
+//
+// There is a single class `trace:trace`
+//
+// # Object
+//
+// A trace object is a JSON `map[string]any` in [NetFlow] format.
+//
+// # Query
+//
+// A query is a [TraceQL] query string, prefixed by `trace:trace:`, for example:
+//
+//	trace:trace:{resource.k8s.namespace.name="tracing-app-k6"}
+//
+// # Store
+//
+// To connect to a tempoStack store use this configuration:
+//
+//	domain: trace
+//	tempostack: URL_OF_TEMPOSTACK_PROXY
+//
+// To connect to plain tempo store use:
+//
+//	domain: trace
+//	tempo: URL_OF_TEMPO
+//
+// [TraceQL]: https://grafana.com/docs/tempo/latest/query/
+//
+// [Trace]: https://docs.openshift.com/container-platform/4.16/observability/distr_tracing/distr_tracing_tempo/distr-tracing-tempo-installing.html
+package trace
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/korrel8r/korrel8r/internal/pkg/tempo"
+	"github.com/korrel8r/korrel8r/pkg/config"
+	"github.com/korrel8r/korrel8r/pkg/domains/k8s"
+	"github.com/korrel8r/korrel8r/pkg/korrel8r"
+	"github.com/korrel8r/korrel8r/pkg/korrel8r/impl"
+)
+
+var (
+	// Verify implementing interfaces.
+	_ korrel8r.Domain = Domain
+	_ korrel8r.Store  = &store{}
+	_ korrel8r.Store  = &stackStore{}
+	_ korrel8r.Query  = Query("")
+	_ korrel8r.Class  = Class{}
+)
+
+// Domain for trace records.
+//
+// There are several possible trace store configurations:
+// - Default TempoStack store on current Openshift cluster: `{}`
+// - Remote TempoStack: `{ "tempoStack": "https://url-of-tempostack"}`
+// - Plain Tempo store: `{ "tempo": "https://url-of-tempo"}`
+var Domain = domain{}
+
+type domain struct{}
+
+func (domain) Name() string                     { return "trace" }
+func (d domain) String() string                 { return d.Name() }
+func (domain) Description() string              { return "Traces from Pods and Nodes." }
+func (domain) Class(name string) korrel8r.Class { return Class{} }
+func (domain) Classes() []korrel8r.Class        { return []korrel8r.Class{Class{}} }
+func (d domain) Query(s string) (korrel8r.Query, error) {
+	_, s, err := impl.ParseQuery(d, s)
+	if err != nil {
+		return nil, err
+	}
+	return Query(s), nil
+}
+
+const (
+	StoreKeyTempo       = "tempo"
+	StoreKeyTempoStack  = "tempoStack"
+	StoreKeyTempoTenant = "tenant"
+)
+
+func (domain) Store(s any) (korrel8r.Store, error) {
+	cs, err := impl.TypeAssert[config.Store](s)
+	if err != nil {
+		return nil, err
+	}
+	hc, err := k8s.NewHTTPClient(cs)
+	if err != nil {
+		return nil, err
+	}
+
+	tempo, tempoStack := cs[StoreKeyTempo], cs[StoreKeyTempoStack]
+	switch {
+
+	case tempo != "" && tempoStack != "":
+		return nil, fmt.Errorf("can't set both tempo and tempoStack URLs")
+
+	case tempo != "":
+		u, err := url.Parse(tempo)
+		if err != nil {
+			return nil, err
+		}
+		return NewPlainTempoStore(u, hc)
+
+	case tempoStack != "":
+		u, err := url.Parse(tempoStack)
+		if err != nil {
+			return nil, err
+		}
+		return NewTempoStackStore(u, hc)
```

Comment:
Instead of `PlainTempoStore` and `TempoStackStore` I'd call it `TempoStore` and `MultiTenantTempoStore` to indicate that the latter uses a multi-tenant tempo deployment with the observatorium gateway